### ISSUE TITLE
[Bug]: 'Map with a query expression' code action does not support for field access expressions

### ISSUE THREAD

**Issue by madushajg** [ballerina-platform/ballerina-lang#38298]
### Description

![2022-10-19 13 29 53](https://user-images.githubusercontent.com/26219651/196865131-9a2d503e-ac53-4167-b87b-18267c716150.gif)

### Steps to Reproduce

Assign an array from a field access expression and find the quick actions

### Affected Version(s)

2201.2.2 RC1

### OS, DB, other environment details and versions

-

### Related area

-> Other Area

**Comment by madushajg** [comment]
I tested this in a custom `2201.4.0` pack and it seems the issue is still there. Maybe I have missed something.
![map-with-query-not-available](https://user-images.githubusercontent.com/26219651/217433338-8a955a7b-9c77-4bd9-9a22-ef7543b80f3a.gif)

@IMS94 @LakshanWeerasinghe Could you please have a look?

**Comment by IMS94** [comment]
Reproducible on master. Reopening

**Comment by malinthar** [comment]
@madushajg @IMS94 
`Map with a query expression` code action is a `range-based` code action and works only when the selected range corresponds to a  syntax node with a syntax kind of  `LOCAL_VARIABLE`, `SPECIFIC_FIELD` or `ASSIGNMENT_STATEMENT`.

In the scenario above there is no range selection. Clicking on the `valueExpr` part of the specific field node makes the  `input.assets` the nodeAtCursor(selected range). `input.assets` is a`FieldAccessNode`. Therefore, not supported by the code action.

In summary, the code action is not diagnostic-based and does not support the FieldAccessNode selection. Therefore the code action is not shown. However,  if you click on the `fieldName` or select the `SpecificFeldNode` the code action is shown as depicted below. 

![Peek 2023-03-07 12-08](https://user-images.githubusercontent.com/35211477/223342560-a7754c38-c7c7-4dc3-804c-0c32815f29cc.gif)

**Comment by madushajg** [comment]
@malinthar Thanks for the update. Yeah, that seems to be working.
But shouldn't both the below cases behave in the same way?

![map-with-query](https://user-images.githubusercontent.com/26219651/223408193-8077f8dd-b28b-41c6-b570-58ac40f4b844.gif)

And in the above one, the second case seems to be `diagnostic-based`. Any reason why it is not the case for the fieldAccessNodes ?

**Comment by malinthar** [comment]
> @malinthar Thanks for the update. Yeah, that seems to be working. But shouldn't both the below cases behave in the same way?
> 
> ![map-with-query](https://user-images.githubusercontent.com/26219651/223408193-8077f8dd-b28b-41c6-b570-58ac40f4b844.gif) [ ![map-with-query](https://user-images.githubusercontent.com/26219651/223408193-8077f8dd-b28b-41c6-b570-58ac40f4b844.gif) ](https://user-images.githubusercontent.com/26219651/223408193-8077f8dd-b28b-41c6-b570-58ac40f4b844.gif) [ ](https://user-images.githubusercontent.com/26219651/223408193-8077f8dd-b28b-41c6-b570-58ac40f4b844.gif)
> 
> And in the above one, the second case seems to be `diagnostic-based`. Any reason why it is not the case for the fieldAccessNodes ?

Yes I agree the UX should be the same. The difference is due to the way we have modelled ranged-based code actions based on the syntax node in the given context. We will address this.
